src/connector/src/sink/big_query.rs

```diff
@@ -583,82 +669,91 @@ impl BigQuerySinkWriter {
-                .await
-            {
-                Ok(_) => {
-                    self.write_rows_count = 0;
-                    self.write_rows.clear();
-                    return Ok(());
-                }
-                Err(e) => errs.push(e),
-            }
-        }
-        Err(SinkError::BigQuery(anyhow::anyhow!(
-            "Insert error {:?}",
-            errs
-        )))
-    }
-}
-
-#[async_trait]
-impl SinkWriter for BigQuerySinkWriter {
-    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+    async fn write_chunk(&mut self, chunk: StreamChunk) -> Result<usize> {
         let serialized_rows = if self.is_append_only {
             self.append_only(chunk)?
         } else {
             self.upsert(chunk)?
         };
-        if !serialized_rows.is_empty() {
-            self.write_rows_count += serialized_rows.len();
+        if serialized_rows.is_empty() {
+            return Ok(0);
+        }
+        let mut result = Vec::new();
+        let mut result_inner = Vec::new();
+        let mut size_count = 0;
+        for i in serialized_rows {
+            size_count += i.len();
+            if size_count > MAX_ROW_SIZE {
+                result.push(result_inner);
+                result_inner = Vec::new();
+                size_count = i.len();
+            }
+            result_inner.push(i);
+        }
+        if !result_inner.is_empty() {
+            result.push(result_inner);
+        }
+        let len = result.len();
+        for serialized_rows in result {
             let rows = AppendRowsRequestRows::ProtoRows(ProtoData {
                 writer_schema: Some(self.writer_pb_schema.clone()),
                 rows: Some(ProtoRows { serialized_rows }),
             });
-            self.write_rows.push(rows);
-
-            if self.write_rows_count >= self.config.common.max_batch_rows {
-                self.write_rows().await?;
-            }
+            self.client
+                .append_rows(rows, self.write_stream.clone())
+                .await?;
         }
-        Ok(())
-    }
-
-    async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
-        Ok(())
-    }
-
-    async fn abort(&mut self) -> Result<()> {
-        Ok(())
+        Ok(len)
     }
+}
 
-    async fn barrier(&mut self, is_checkpoint: bool) -> Result<()> {
-        if is_checkpoint {
-            self.write_rows().await?;
+#[try_stream(ok = (), error = SinkError)]
+pub async fn resp_to_stream(
+    resp_stream: impl Future<
+            Output = std::result::Result<
+                Response<google_cloud_gax::grpc::Streaming<AppendRowsResponse>>,
+                Status,
+            >,
+        >
+        + 'static
+        + Send,
+) {
+    let mut resp_stream = resp_stream
+        .await
+        .map_err(|e| SinkError::BigQuery(e.into()))?
+        .into_inner();
+    loop {
+        if let Some(append_rows_response) = resp_stream
```

Comment:
When we reach the end of the `resp_stream`, we will keep yielding (). This is not correct. Instead of using `if let Some(...)`, we'd better just `resp_stream.message().await.ok_or_else(|| ...)` to turn `None` into an end-of-stream error.

Reply:
It doesn't return () at the end of the stream, after receiving every reply, if there are no errors, it returns a meaningless (), which calls the method of the stream, which reaches the end of the stream.

Reply:
I know the purpose of returning a meaningless `()`. What I meant was that, when `resp_stream.message()` returns None, we should break the loop with an `EndOfStream` error, or just simply break the loop, instead of ignoring it. 

In current code, after resp_stream returns None, an `()` will still be yielded, and later when polled again, and resp_stream returns None again, and an `()` is still yielded, and this will be repeated endlessly over and over again, while the external code is not aware that the response stream has actually stopped.